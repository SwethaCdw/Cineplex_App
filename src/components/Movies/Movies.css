.movies-list-section {
  display: flex;
  gap: 150px;
}
.movies-list-section .all-movies {
  flex: 40%;
}
.movies-list-section .all-movies .load-more-btn {
  background-color: #d9bc2d;
  padding: 10px 20px;
  color: white;
  border: none;
  margin-left: 35px;
}
.movies-list-section .all-movies .all-movies-title {
  margin-left: 40px;
  margin-top: 30px;
  font-size: 35px;
}
.movies-list-section .all-movies .movie-list {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 20px;
  margin: 20px 35px;
}
.movies-list-section .all-movies .movie-list .movie {
  padding: 5px;
}
.movies-list-section .all-movies .movie-list .movie img {
  width: 220px;
  height: 300px;
  -o-object-fit: cover;
     object-fit: cover;
  cursor: pointer;
}
.movies-list-section .all-movies .movie-list .movie .movie-details-container {
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
}
.movies-list-section .movies-description {
  flex: 60%;
  margin-top: 90px;
}
.movies-list-section .movies-description .selected-movie-details {
  margin-right: 40px;
}
.movies-list-section .movies-description .selected-movie-details .selected-movie {
  display: flex;
  align-items: center;
  justify-content: space-between;
}
.movies-list-section .movies-description .selected-movie-details .selected-movie-name {
  font-size: 35px;
}
.movies-list-section .movies-description .selected-movie-details .selected-movie-likes {
  margin-bottom: 25px;
}
.movies-list-section .movies-description .selected-movie-details .selected-movie-poster {
  width: 200px;
  height: 250px;
  -o-object-fit: cover;
     object-fit: cover;
}
.movies-list-section .movies-description .selected-movie-details .selected-movie-description {
  line-height: 1.5;
  margin-top: 20px;
  font-size: 18px;
}
.movies-list-section .movies-description .selected-movie-details .actors-title {
  margin: 25px 0;
}
.movies-list-section .movies-description .selected-movie-details .actors-list {
  line-height: 1.5;
  list-style: none;
}
.movies-list-section .movies-description .selected-movie-details .actors-list .actor {
  font-size: 18px;
}
.movies-list-section .notification {
  margin-right: 40px;
  margin-top: 20px;
  font-size: 20px;
  color: #488cd7;
}
.movies-list-section .notification .advertisement {
  width: 800px;
  height: 700px;
  -o-object-fit: contain;
     object-fit: contain;
}
.movies-list-section .thumbs-up-button {
  background: none;
  border: none;
  cursor: pointer;
  padding: 0;
}
.movies-list-section .thumbs-up-button svg {
  transition: transform 0.3s ease;
}
@keyframes bounce {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.3);
  }
  100% {
    transform: scale(1);
  }
}
.movies-list-section .thumbs-up-button.liked {
  animation: bounce 0.5s ease;
  color: red;
}/*# sourceMappingURL=Movies.css.map */