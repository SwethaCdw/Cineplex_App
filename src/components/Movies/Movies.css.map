{"version":3,"sources":["Movies.scss","Movies.css","../../styles/colors.scss","../../styles/mixins.scss"],"names":[],"mappings":"AAKA;EACI,aAAA;EACA,UAAA;ACJJ;ADKI;EACI,SAAA;ACHR;ADIQ;EACI,yBELW;EFMX,kBAAA;EACA,WEJC;EFKD,YAAA;EACA,iBAAA;ACFZ;ADIQ;EACI,iBAAA;EACA,gBAAA;EACA,eAAA;ACFZ;ADIQ;EACI,aAAA;EACA,qCAAA;EACA,cAAA;EACA,iBAAA;EACA,YAAA;EACA,cAAA;ACFZ;ADGY;EACI,YAAA;ACDhB;ADEgB;EACI,YAAA;EACA,aAAA;EACA,oBAAA;KAAA,iBAAA;EACA,eAAA;ACApB;ADEgB;EGpCZ,aAAA;EACA,uBHoCmD;EGnCnD,8BHmCoC;ACExC;ADDoB;EACI,UAAA;ACGxB;ADII;EACI,SAAA;EACA,gBAAA;ACFR;ADGQ;EACI,kBAAA;ACDZ;ADEY;EGnDR,aAAA;EACA,mBHmD+C;EGlD/C,8BHkDgC;ACEpC;ADAY;EACI,eAAA;ACEhB;ADAY;EACI,cAAA;EACA,cAAA;EACA,gBAAA;ACEhB;ADAY;EACI,YAAA;EACA,aAAA;EACA,oBAAA;KAAA,iBAAA;ACEhB;ADAY;EACI,gBAAA;EACA,gBAAA;EACA,eAAA;ACEhB;ADAY;EACI,cAAA;ACEhB;ADAY;EACI,gBAAA;EACA,gBAAA;ACEhB;ADDgB;EACI,eAAA;ACGpB;ADEI;EACI,kBAAA;EACA,gBAAA;EACA,eAAA;EACA,cEvFU;ADuFlB;ADCQ;EACI,YAAA;ACCZ;ADAY;EACI,WAAA;EACA,aAAA;EACA,oBAAA;KAAA,iBAAA;ACEhB;ADGI;EACI,WAAA;EACA,YAAA;EACA,gBAAA;EACA,eAAA;EACA,UAAA;ACDR;ADGQ;EACI,uBAAA;EACA,kBAAA;ACDZ;ADIQ;EACI,+BAAA;ACFZ;ADIY;EACI;IACI,mBAAA;ECFlB;EDIc;IACI,qBAAA;ECFlB;EDIc;IACI,mBAAA;ECFlB;AACF;ADOQ;EACI,2BAAA;EACA,YE3HE;ADsHd","file":"Movies.css"}