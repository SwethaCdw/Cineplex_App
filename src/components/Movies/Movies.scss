
@import '../../styles/variables.scss';
@import '../../styles/colors.scss';
@import '../../styles/mixins.scss';

.movies-list-section {
    display: flex;
    gap: 150px;
    .all-movies {
        flex: 40%;
        .load-more-btn {
            background-color: $primary-button-color;
            padding: 10px 20px;
            color: $text-color;
            border: none;
            margin-left: 35px;
        }
        .all-movies-title {
            margin-left: 40px;
            margin-top: 30px;
            font-size: 35px;
        }
        .movie-list {
            display: grid;
            grid-template-columns: repeat(3, 1fr);
            grid-gap: 20px;
            margin: 20px 35px;
            height: 75vh;
            overflow: auto;
            .movie {
                padding: 5px;
                img{
                    width: 220px;
                    height: 300px;
                    object-fit: cover;
                    cursor: pointer;
                }
                .movie-details-container {
                    @include flex-align(space-between, flex-start);
                    .movie-details {
                        width: 80%;
                    }
                }
            }
          
        }
    }
    .movies-description {
        flex: 60%;
        margin-top: 90px;
        .selected-movie-details {
            margin-right: 40px;
            .selected-movie {
                @include flex-align(space-between, center);
            }
            .selected-movie-name {
                font-size: 35px;
            }
            .selected-movie-likes {
                margin: 18px 0;
                color: #d9bc2d;
                font-weight: 900;
            }
            .selected-movie-poster {
                width: 200px;
                height: 250px;
                object-fit: cover;
            }
            .selected-movie-description {
                line-height: 1.5;
                margin-top: 20px;
                font-size: 18px;
            }
            .actors-title {
                margin: 25px 0;
            }
            .actors-list {
                line-height: 1.5;
                list-style: none;
                .actor {
                    font-size: 18px;
                }
            }
        }
    }
    .notification {
        margin-right: 40px;
        margin-top: 20px;
        font-size: 20px;
        color: $secondary-color;
        .ad-notify-container {
            width: 720px;
            .advertisement {
                width: 100%;
                height: 700px;
                object-fit: cover;
            }
        }
       
    }
    .thumbs-up-button {
        width: 40px;
        height: 40px;
        background: none;
        cursor: pointer;
        padding: 0;

        &:hover, &.description {
            border: 1px solid white;
            border-radius: 50%;
        }
    
        svg {
            transition: transform 0.3s ease;
    
            @keyframes bounce {
                0% {
                    transform: scale(1);
                }
                50% {
                    transform: scale(1.3);
                }
                100% {
                    transform: scale(1);
                }
            }
    
        
        }
        &.liked {
            animation: bounce 0.5s ease;
            color: $liked-color;
        }
    }
}