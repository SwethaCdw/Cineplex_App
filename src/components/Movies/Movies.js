import React, { useEffect, useState } from 'react';
import './Movies.css';
import { getMovies } from '../../services/getMovies';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faThumbsUp } from '@fortawesome/free-solid-svg-icons';
import withAdvertisement from '../withAdvertisement/withAdvertisement';
import  AdvImage1  from '../../assets/advertisements/large-promos/adv1.png';
import  AdvImage2  from '../../assets/advertisements/large-promos/adv-2.png';
import { formatCounter, getRandomItem, handleImageError } from '../../utils/common-utils';
import { ACTORS, LIKES, LOAD_MORE } from '../../constants/common-constants';
import { ADVERTISEMENT_IN, DESCRIPTION_COUNTER, RESUMES_IN, TITLE, VISIBLE_MOVIES_COUNTER } from '../../constants/movie-constants';
import Button from '../../common/Button/Button';
import Image from '../../common/Image/Image';

const Movies = React.memo(({ moviePageCounter, showImage, imageCounter, setMoviePageCounter }) => {
    console.log('COMPONENT :: Movies')
    const [selectedMovie, setSelectedMovie] = useState(null);
    const [movies, setMovies] = useState([]);
    const [visibleMovies, setVisibleMovies] = useState(VISIBLE_MOVIES_COUNTER); 
    const [totalMoviesLength, setTotalMoviesLength] = useState(0);
    const [likedMovies, setLikedMovies] = useState([]); 
    const [randomAdvertisementImage, setRandomAdvertisementImage] = useState('');

    /**
     * Handle Like click
     * @param {*} movieName 
     */
    const handleLikeClick = (movieName) => {
        const updatedMovies = [...movies];
        const movieToUpdate = updatedMovies.find(movie => movie.movie_name === movieName);
    
        if (movieToUpdate) {
            if (likedMovies.includes(movieName)) {
                movieToUpdate.movie_likes -= 1;
                setLikedMovies(likedMovies.filter(name => name !== movieName));
            } else {
                movieToUpdate.movie_likes += 1;
                setLikedMovies([...likedMovies, movieName]);
            }
            setMovies(updatedMovies);
        }
    };
    
    /**
     * Handle movie click
     * @param {*} movie 
     */
    const handleMovieClick = (movie) => {
        setSelectedMovie(movie);
        setMoviePageCounter([DESCRIPTION_COUNTER]);      
    };

    useEffect(() => {
        setMoviePageCounter([DESCRIPTION_COUNTER]); 
        fetchMovies();
        const advertisementImages = [AdvImage1, AdvImage2];
        const adImage = getRandomItem(advertisementImages);
        setRandomAdvertisementImage(adImage);
    }, []);

    /**
     * Fetch movies from API
     */
    const fetchMovies = async () => {
        try {
            const response = await getMovies(visibleMovies);
            if (!response || !response.movies) {
                throw new Error('Movies not found in API response');
            }
            const { movies, totalLength } = response;
            setMovies(movies);
            setSelectedMovie(movies[0]);
            setTotalMoviesLength(totalLength);
        } catch (error) {
            console.error('Error fetching movies:', error);
        }
    };

    /**
     * Load more movies on click of Load more button
     */
    const loadMoreMovies = async () => {
        try {
            const { movies } = await getMovies(visibleMovies + VISIBLE_MOVIES_COUNTER); // Fetch additional movies
            setMovies(movies);
            setVisibleMovies(prevVisibleMovies => prevVisibleMovies + VISIBLE_MOVIES_COUNTER); // Increase the number of visible movies
        } catch (error) {
            console.error('Error loading more movies:', error);
        }
    };

    return (
        <div className='movies-list-section'>
            <section className='all-movies'>
                <p className='all-movies-title'>{TITLE}</p>
                <div className='movie-list'>
                    {movies.map((movie, index) => (
                        <div className='movie' key={index}>  
                            <Image  src={movie.movie_poster} alt={movie.movie_name} onError={handleImageError} onClick={() => handleMovieClick(movie)}/>
                            <div className='movie-details-container'>
                                <div className='movie-details'>
                                    <p data-testid='movie-name' className='movie-name'>{movie.movie_name}</p>
                                    <p className='movie-likes'>{movie.movie_likes} {LIKES}</p>
                                </div>
                                <Button  className={`thumbs-up-button ${likedMovies.includes(movie.movie_name) ? 'liked' : ''}`} onClick={() => handleLikeClick(movie.movie_name)} children={ <FontAwesomeIcon icon={faThumbsUp} />} />
                            </div>
                        </div>
                    ))}
                </div>
                {totalMoviesLength > visibleMovies && (<Button className='load-more-btn' onClick={loadMoreMovies} children={LOAD_MORE}></Button> )}

                
            </section>
            <section className='movies-description'>
                {(selectedMovie && !showImage[0] )&& (
                    <div className='selected-movie-details'>
                        <div className='selected-movie'>
                            <h2 className='selected-movie-name'>{selectedMovie.movie_name}</h2>
                            <Button data-testid='like-button'  className={`thumbs-up-button description ${likedMovies.includes(selectedMovie.movie_name) ? 'liked' : ''}`} onClick={() => handleLikeClick(selectedMovie.movie_name)} children={ <FontAwesomeIcon icon={faThumbsUp} />}  />
                        </div>
                        <p className='selected-movie-likes'>{selectedMovie.movie_likes} {LIKES}</p>
                        <Image className='selected-movie-poster' src={selectedMovie.movie_poster} alt={selectedMovie.movie_name} onError={handleImageError} />
                        <p className='selected-movie-description'>{selectedMovie.movie_description}</p>
                        <h2 className='actors-title'>{ACTORS}</h2>
                        <ul className='actors-list'>
                            {selectedMovie.movie_actors.map((actor, index) => (
                                <li className='actor' key={index}>{actor}</li>
                            ))}
                        </ul>
                     
                    </div>
                  
                )}
                <div className='notification'>
                    {showImage[0] ? (
                        <div className='ad-notify-container'>
                        <Image src={randomAdvertisementImage} alt='advertisement' className='advertisement' onError={handleImageError}/>
                        {imageCounter > 0 && <p>{RESUMES_IN} {formatCounter(imageCounter)}</p>}
                        </div>
                        ) : (
                        moviePageCounter > 0 && <p>{ADVERTISEMENT_IN} {formatCounter(moviePageCounter)}</p>
                        )}
                </div> 
            </section>
        </div>
    );
});
export default withAdvertisement(Movies, false);