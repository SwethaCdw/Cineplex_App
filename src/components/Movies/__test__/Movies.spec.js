import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import Movies from '../Movies';

test('load movies', () => {
    const { asFragment } = render(<Movies />);
    expect(asFragment()).toMatchSnapshot();
      
})

test('is Movie title present', async () => {
    render(<Movies />);
    const MovieText = screen.getByText('All Movies');
    expect(MovieText).toBeInTheDocument();
});

// jest.mock('../../../services/getMovies', () => ({
//     getMovies: jest.fn().mockResolvedValue({
//       movies: [
//         {
//           movie_name: 'Test Movie 1',
//           movie_poster: 'test-poster-1.jpg',
//           movie_description: 'Description of Test Movie 1',
//           movie_likes: 100,
//           movie_actors: ['Actor 1', 'Actor 2']
//         },
//         {
//           movie_name: 'Test Movie 2',
//           movie_poster: 'test-poster-2.jpg',
//           movie_description: 'Description of Test Movie 2',
//           movie_likes: 200,
//           movie_actors: ['Actor 3', 'Actor 4']
//         }
//       ],
//       totalLength: 2
//     })
//   }));


//   test('likes a movie on button click', async () => {
//     render(<Movies />);
//     const likeButton = await screen.findByRole('like-button', { name: 'Test Movie 1' });
//     fireEvent.click(likeButton);
//     expect(likeButton).toHaveClass('liked');
//   });

//   test('displays movie details on click', async () => {
//     render(<Movies />);
//     const moviePoster = await screen.findByAltText('Test Movie 1');
//     fireEvent.click(moviePoster);
//     const movieName = await screen.findByText('Test Movie 1');
//     expect(movieName).toBeInTheDocument();
//   });

//   test('loads more movies on button click', async () => {
//     render(<Movies />);
//     const loadMoreButton = await screen.findByText('Test Movie 1');
//     fireEvent.click(loadMoreButton);
//     await waitFor(() => {
//       expect(screen.getByAltText('Test Movie 2')).toBeInTheDocument();
//     });
//   });


test('changes background color to red when like button is clicked', async () => {
    render(<Movies />);

    await waitFor(() => {
        expect(screen.getByTestId('like-button')).toBeInTheDocument();
    });

    const likeButton = screen.getByTestId("like-button");
    const initialBackgroundColor = window.getComputedStyle(likeButton).getPropertyValue('background-color');
  
    // Click the like button
    fireEvent.click(likeButton);
  
    const updatedBackgroundColor = window.getComputedStyle(likeButton).getPropertyValue('background-color');
  
    expect(initialBackgroundColor).not.toBe(updatedBackgroundColor);
    expect(updatedBackgroundColor).toBe('rgb(255, 0, 0)'); 
  });