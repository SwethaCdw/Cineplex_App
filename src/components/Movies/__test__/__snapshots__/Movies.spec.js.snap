// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load movies 1`] = `
<DocumentFragment>
  <div
    class="movies-list-section"
  >
    <section
      class="all-movies"
    >
      <p
        class="all-movies-title"
      >
        All Movies
      </p>
      <div
        class="movie-list"
      />
    </section>
    <section
      class="movies-description"
    >
      <div
        class="notification"
      >
        <p>
          Advertisement in 0:15
        </p>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
